ui_test/src/lib.rs

```diff
@@ -0,0 +1,414 @@
+    );
+    // Check error annotations in the source against output
+    check_annotations(&output.stderr, &mut errors, config, revision, comments);
+    (miri, errors)
+}
+
+fn check_annotations(
+    unnormalized_stderr: &[u8],
+    errors: &mut Errors,
+    config: &Config,
+    revision: &str,
+    comments: &Comments,
+) {
+    let unnormalized_stderr = std::str::from_utf8(unnormalized_stderr).unwrap();
+    let mut found_annotation = false;
+    if let Some((ref error_pattern, definition_line)) = comments.error_pattern {
+        if !unnormalized_stderr.contains(error_pattern) {
+            errors.push(Error::PatternNotFound {
+                stderr: unnormalized_stderr.to_string(),
+                pattern: error_pattern.to_string(),
+                definition_line,
+            });
+        }
+        found_annotation = true;
+    }
+    for &ErrorMatch { ref matched, revision: ref rev, definition_line } in &comments.error_matches {
+        // FIXME: check that the error happens on the marked line
+
+        if let Some(rev) = rev {
+            if rev != revision {
+                continue;
+            }
+        }
+
+        if !unnormalized_stderr.contains(matched) {
+            errors.push(Error::PatternNotFound {
+                stderr: unnormalized_stderr.to_string(),
+                pattern: matched.to_string(),
+                definition_line,
+            });
+        }
+        found_annotation = true;
+    }
+    match (config.mode, found_annotation) {
+        (Mode::Pass, true) | (Mode::Panic, true) => errors.push(Error::PatternFoundInPassTest),
+        (Mode::Fail, false) => errors.push(Error::NoPatternsFound),
+        _ => {}
+    };
+}
+
+fn check_output(
+    output: &[u8],
+    path: &Path,
+    errors: &mut Errors,
+    kind: String,
+    target: &str,
+    filters: &Filter,
+    config: &Config,
+    comments: &Comments,
+) {
+    let output = std::str::from_utf8(&output).unwrap();
+    let output = normalize(path, output, filters, comments);
+    let path = output_path(path, comments, kind, target);
+    if config.bless {
+        if output.is_empty() {
+            let _ = std::fs::remove_file(path);
+        } else {
+            std::fs::write(path, &output).unwrap();
+        }
+    } else {
+        let expected_output = std::fs::read_to_string(&path).unwrap_or_default();
+        if !config.skip_output_checks {
+            if output != expected_output {
+                errors.push(Error::OutputDiffers {
+                    path,
+                    actual: output,
+                    expected: expected_output,
+                });
+            }
+        }
+    }
+}
+
+fn output_path(path: &Path, comments: &Comments, kind: String, target: &str) -> PathBuf {
+    if comments.stderr_per_bitwidth {
+        return path.with_extension(format!("{}.{kind}", get_pointer_width(target)));
+    }
+    path.with_extension(kind)
+}
+
+fn compare_output(path: &Path, actual: &str, expected: &str) {
```

Comment:
This is only called when the output differs and we are printing that error, right?
Please pick a function name that better reflects that, or add a doc comment, or both.
Also the "if actual == expectd" seems off then?